### PR TITLE
No flatten for most backends

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
Do not flatten matches for most backends. Such flattening can cause the generated code size to be the square of the input code size. Removing the flattening leads to a huge decrease in the size of the generated C# code inside the Dafny codebase. It goes from ~14K lines to ~7K.

The compilation of nested matches is done as follows:
```dafny
datatype TX = X(x1arg: TY, x2arg: TY)
datatype TY = Y(yarg: int)
datatype TZ = Z(zarg: TW)
datatype TW = W(warg: int)

method M(a: TX) {
  match a
    case X(Y(b),Z(W(c)) => <body1>
    case r => <body2>
}
```

Is roughly compiled into 

```dafny
// Same datatypes

method M(a: TX) {
  var unmatched := true;
  if (unmatched && a is X) {
    var x1arg1 = ((X)a).1;
    if (x1 is Y) {
      var b = ((Y)x1arg1).1;
      var x2arg2 := ((X)a).2; 
      if (x2 is Z) {
        var zarg1 := ((Z)x2arg2).1;
        if (x4 is W) {
          var c := ((W)zarg1).1;
          unmatched := false;
          <body1>
        }
      } 
    }
  }
  if (unmatched) {
    var r := a;
    <body2>
  }
}
```

##### Maintainability 
To reduce the required work, Java and Dafny back-ends still compile using flattened matches

Ideally the transformation would be a Dafny-to-Dafny source transformation, instead of a customization of SinglePassCompiler. However, Dafny does not allow using statements in expression contexts, and this is needed for the transformation. I think it would be good to have an intermediate Dafny that does allow this, similar to what @cpitclaudel 's Dafny-in-Dafny compiler allowed, and then to define the rewrite that this PR introduces using a Dafny source translation.

##### Improvement
For C# we could generate much nicer code, since C# allows declaring new variables inside expressions using `x is T xAsType` expressions. We could get rid of the nested `if` statements and the `unmatched` variable. However, I'll leave this for future work.

### How has this been tested?
- Performance change. No additional tests added.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Comment by robin-aws** [comment]
Just for my own understanding, why did the original code generation lead to only a quadratic increase in code size? I would think that it would be a factor of the number of other data constructors, and be worse than quadratic for more nested matches, i.e. if the example above had lots of data constructors for each datatype, I'd expect the worst case to be something like `O(|TX| * |TZ| * |TW|)`

**Comment by keyboardDrummer** [comment]
> Just for my own understanding, why did the original code generation lead to only a quadratic increase in code size? I would think that it would be a factor of the number of other data constructors, and be worse than quadratic for more nested matches, i.e. if the example above had lots of data constructors for each datatype, I'd expect the worst case to be something like `O(|TX| * |TZ| * |TW|)`

I guess if your nested match occurs in a case that gets copied, like a default case, then it could be exponential yes.

But if your nested match occurs in a case that does not get copied, then the nesting has no particular effect.

However, regardless of nesting, for each match in the original program, the amount of flattened cases will be the number of constructors of the datatype, so the size will be `#matches` * `#ofConstructors` for a program where all matches match on the same datatype and all constructor are part of that datatype.

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs [hunk 1]
```diff
@@ -11,8 +18,6 @@
 
 namespace Microsoft.Dafny.Compilers {
   public abstract partial class SinglePassCodeGenerator {
-
-
     protected void TrStmt(Statement stmt, ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts = null) {
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs:21]: Applied a resharper refactoring to turn the if's into a switch statement. No actual changes to this method except for NestedMatchStmt, which now calls into a method.

File: Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs [hunk 2]
```diff
@@ -460,5 +509,162 @@ public abstract partial class SinglePassCodeGenerator {
         }
       }
     }
+
+    protected virtual void EmitNestedMatchStmt(NestedMatchStmt match, ConcreteSyntaxTree writer) {
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.Statement.cs:513]: This is where the changes in this file start.

**Review by MikaelMayer — APPROVED** [review]
This looks good to me.

**Pull request merged**